core/src/main/scala/org/bitcoins/core/wallet/utxo/AddressTag.scala

```diff
@@ -0,0 +1,74 @@
+package org.bitcoins.core.wallet.utxo
+
+sealed trait AddressTagType {
+  def tagType: String
+}
+
+sealed trait AddressTag {
+  def name: String
+  def tagType: String
+
+  def ==(at: AddressTag): Boolean = name == at.name && tagType == at.tagType
+  def !=(at: AddressTag): Boolean = !(this == at)
+
+  def toCustom: CustomAddressTag = CustomAddressTag(name, tagType)
+}
+
+trait AddressTagFactory[TagType <: AddressTagType] {
+
+  def tagType: String
+
+  def all: Vector[TagType]
+
+  def fromString(str: String): Option[TagType] = {
+    all.find(tag => str.toLowerCase() == tag.toString.toLowerCase)
+  }
+
+  def fromCustom(custom: CustomAddressTag): Option[TagType] = {
+    // Allow empty tagType because sometimes (like in the db mappers) we don't have access to
+    // the tagType when we are getting the tag, so we need to be able to pass it in.
+    require(
+      custom.tagType.isEmpty || custom.tagType == tagType,
+      s"If tag's tagType is defined it must be the same as tagType, got ${custom.tagType}")
+
+    fromString(custom.name)
+  }
+}
+
+sealed trait InternalAddressTagType extends AddressTagType
+sealed trait InternalAddressTag extends AddressTag
+
+trait ExternalAddressTagType extends AddressTagType
+trait ExternalAddressTag extends AddressTag
```

Comment:
This won't work, `sealed` is inherited (otherwise it wouldn't work), so you're going to have to make the top type not `sealed` (and I would suggest it go in another file). And only `InternalAddressTag[Type]` will be `sealed`